system/ofloader/segger.c

```diff
@@ -0,0 +1,244 @@
+/****************************************************************************
+ * apps/system/ofloader/segger.c
```

Comment:
shouldn't this file be on the kernel side? direct call to `__start` doesn't seem appropriate in apps

Reply:
Yes, it isn't good to call __start from app, but since ofloader is a very low-level tool which interact JLINK directly:

1. JLINK require flash programmer must start with SEGGER_FL_Prepare in PrgData section
2. ofloader  is designed to work only in FLAT mode, it can work __start without trouble

Do you have the better ideal?

Reply:
Maybe it's worth moving the whole `ofloader` to the kernel side ? Is it foreseen that the `ofloader` will be run from NSH? or only as an entry point?

If we want to keep it in apps then we should use `boardctl` interface to call `__start`.

Reply:
> Maybe it's worth moving the whole `ofloader` to the kernel side ? Is it foreseen that the `ofloader` will be run from NSH? or only as an entry point?

Yes, it's better to run from NSH since we want to support multiple flash method(e.g. ymodem, fastboot and ofloader) at the same time.

> 
> If we want to keep it in apps then we should use `boardctl` interface to call `__start`.

Here is how it work under JLINK:

1. JLINK load ofloader to SRAM through JTAG
2. JLINK jump PC to SEGGER_FL_Prepare 
3. SEGGER_FL_Prepare initialize the chip and return
4. JLINK prepare the data and jump PC to SEGGER_FL_Program
5. SEGGER_FL_Program write a chunk of data to flash and return

Repeat step 4-5 until the whole image is written to flash. The whole flash is controlled by JLINK, and boardctl can't work here, since __start can only run once.

Reply:
I'm not going to block this PR but let's wait a little with the merge, maybe someone will have some valuable feedback. I'm not familiar with this tool from Segger and I'm unable to play with it at the moment.

Reply:
Sure, ofloader is designed to work in bootloader, and most bootloader is built with flat mode to make it as simple as possible, so the limitation here isn't too rigid as the first glance.

Reply:
Yes, we have tried our best to make the violation as minimal as possible. This is a best solution we got after @anjiahao1 try several possible approaches.